### DISCUSSION TITLE
[RFC] User API for integrators

### DISCUSSION THREAD

**Discussion by mociepka** [saleor/saleor#11800] [RFC]
As a saleor integrator, I need an API that I can use to import/synchronize users from different systems.

Right now, we can create user by the `accountRegister` mutation, but It requires a password in plain text format.
This API is great for user registration, not for importing existing accounts because other systems do not save passwords in plain text (god bless).

I propose adding an API that allows integrators to create and update users with password hashes. API that is available only with special permissions and **has a bulk version** (API in which I can send thousands of users at once).

Django already has a great approach for password hashes standard:
https://docs.djangoproject.com/en/4.1/topics/auth/passwords/#how-django-stores-passwords

### New user API
I propose to reuse Django standard with mutations like `createUser`, `updateUser` with input like:
```SDL
input UserInput {
  firstName: String
  lastName: String
  email: String!

  ...
 
  passwordAlgorithm: passwordAlgorithmEnum - from existing password hashers
  passwortHashIterations: Int
  passwordSalt: String
  passwordHash: String
}
```

### Ability to add new password hashers
API above will work with hashes used [currently in Saleor](https://github.com/saleor/saleor/blob/main/saleor/settings.py#L197). For some of the clients, that list will need to be longer.

I propose to give Saleor enterprise clients the ability to add new hashers. Restrictions agreed with @patrys:

- hashers need to use salt (to prevent rainbow table vector of deencription),
- if client hashes don't contain salt (ex., only SHA1), the hash should be wrapped by `pbkdf2` algorithm ([more info in the Django site](https://docs.djangoproject.com/en/4.1/topics/auth/passwords/#password-upgrading-without-requiring-a-login)) and this hasher can be added to the Saleor. 
- hashers need to use existing algorithms, no custom builds per client (in that case, OpenID Connect protocol needs to be used).

**Comment by patrys** [comment]
I don't think Saleor needs to deal with splitting algo, iterations, etc. I'd prefer we only take the password hash as a string and verify the algorithm is supported.

**Reply by mociepka** [reply]
I thought a list of avaiable algorithms as an Enum would be helpful. So you are proposing something like:

```
input UserInput {
  firstName: String
  lastName: String
  email: String!

  ...
  passwordHash: String
}
```

**Comment by piotrjmm** [comment]
I want to add our current use case and problem. We are working on the Open ID Connect Provider solution which we will use for client authorization and we will configure it in Saelor plugin. We have an old system and a new system based on Saleor. During some time period, both systems will work in a way that in some countries, we will use the old system and in some, we will use Saleor. And the problem is that we want to add users during the login process in our provider (there we will authenticate them from the old system's database) and migrate them to Saleor step by step. So we need to do something like accountRegister because we will have the username and password checked by the old database to do it instead of migrating hashes. But the main problem is that this mutation doesn’t have proper input to switch off activation by email. It is only possible to switch it off in environment settings. We need something like this input with an additional field

```graphql
type AccountRegisterInput {
  firstName: String
  lastName: String
  languageCode: LanguageCodeEnum
  email: String!
  password: String!
  redirectUrl: String
  metadata: [MetadataInput!]
  channel: String
}
```


And this field could be it:
requiresConfirmation: Boolean -> it is what is now only returned as information if confirmation is required but it could be used as input parameter

It will make our job easier :)

**Comment by rafalp** [comment]
So there are two problems to solve, and both start with _legacy users database_.

@mociepka wants to pull those users from old database and push them to Saleor with GraphQL API. Their passwords hashes are available, and would be nice to migrate too, so there should be a way to create user together with password hash in format understood by Django.

@piotrjmm has middleware for user authentication. If user is not found in saleor, their credentials are used against old system instead. So he has sign-in credentials (including plaintext password) and user data from legacy system. He now wants to use this data to create new account. `accountRegister` would work but it implements new user registration flow which is invalid for existing users.

Saleor already has following mutations for integrations to create/update existing users in Saleor's database:

- `customerCreate`
- `customerUpdate`
- `staffCreate`
- `staffUpdate`

Both scenarios are about creating new staff/customer account from old user data, so we can focus on `customerCreate` and `staffCreate` mutations only. `customerCreate` requires `MANAGE_USERS` permission and `staffCreate` requires `MANAGE_STAFF` permission.

So integrator knows from old database if user is either staff or customer, so they can use either of those two mutations to create valid user.

The way I see it we only need to add two fields to `UserCreateInput` and `StaffCreateInput`:

```graphql
input CreateInput {
  password: String
  passwordHash: String
}
```

Currently there's `redirectUrl: String` field on inputs which triggers Saleor to direct user to a view where they should setup password for themselves.

With those two extra input logic for customer/staff creation is expanded:

- There are `password`, `passwordHash` and `redirectUrl` fields. If input has more than one of those, it's a validation error.
- If input has `password` value, we hash that and set is as user's password. Do we want to run extra validation for those passwords? It may prevent users with dangerous legacy passwords from registering, which may not be welcome.
- If input has `passwordHash` value, we validate that its [valid Django password hash](https://docs.djangoproject.com/en/4.1/topics/auth/passwords/#how-django-stores-passwords) and then save it in database verbatim.

When `password` or `passwordHash` is used to create an account, no e-mail is sent to customer. Reasoning for this is to make migration transparent for the user.

Is this solution acceptable? @patrys @mociepka @piotrjmm

**Comment by rafalp** [comment]
@piotrjmm I've had a call with @mociepka and @patrys today during which we've realized that `passwordChange` mutation will potentially be more useful to the problem than supporting password migration through `xyzCreate` mutations.

Proposed flow would be this:

- User logins using the `OIC`
- User enters e-mail and password which are verified with old system
- User's account (without password) is created by `OIC` in Saleor
- User is logged in to Saleor
- Login app calls `passwordChange(newPassword: ...)` mutation as `logged in user` with known user's password verified against old system.

Currently `passwordChange` requires that user has password set, but we can change it to support setting first password when user was created without any password.

**Comment by piotrjmm** [comment]
Thank you @rafalp for your response, both solutions are ok for me. When do we expect it to be ready?

**Reply by rafalp** [reply]
PR: https://github.com/saleor/saleor/pull/11999

**Reply by rafalp** [reply]
@piotrjmm `passwordChange` has been merged. Please let us know how it works for you.

**Reply by piotrjmm** [reply]
I just tested it and I can answer, I am not sure about flow because as you mention flow above user login by OIDC after the correct authentication account is created but if you mean by OIDC I understand it as customerCreate or something but an account is created by Saleor from OIDC data.  And now problem is that to change the password by this mutation old password should be set as unusable but is not. And I get an error from this part of the code: 
`            if user.has_usable_password():
                cls.raise_invalid_credentials()`

I am not sure if I miss something or in some step in Saleor unusable password should be set and currently is not?

**Reply by patrys** [reply]
So you're saying that OIDC is creating user accounts with usable passwords? That looks like a bug.

**Reply by piotrjmm** [reply]
So there will be some fix? Or you will check it?

**Reply by patrys** [reply]
Can you answer my question?

**Reply by piotrjmm** [reply]
Ok, sorry :) I think so because I make some logs in Saleor app in passwordChange mutation and this check user.has_usable_password() return true and user is created by OIDC flow.

**Reply by piotrjmm** [reply]
maybe something like this should be added in plugins/openid_connect/utils.py (line near 336):

```
    try:
        user = User.objects.get(**get_kwargs)
    except User.DoesNotExist:
        user, _ = User.objects.get_or_create(
            email=user_email,
            defaults=defaults_create,
        )
        user.set_unusable_password()
        user.save()
    except User.MultipleObjectsReturned:
        logger.warning("Multiple users returned for single OIDC sub ID")
        user, _ = User.objects.get_or_create(
            email=user_email,
            defaults=defaults_create,
        )
        user.set_unusable_password()
        user.save()
```


This change works correctly, I am not sure, but there are probably more places, for example, in user_authorization_flow, but in my case, it is enough without legacy flow.

**Reply by rafalp** [reply]
Other solution that comes to mind is adding `"password": make_password(None),` to `defaults_create` dict that's only used when new user is created.

**Reply by patrys** [reply]
Looks like we also need to fix the password field for accounts that already exist and have an empty string there.

**Reply by rafalp** [reply]
Ok. I'll look at this.

**Reply by rafalp** [reply]
PR with fix: https://github.com/saleor/saleor/pull/12103

**Reply by piotrjmm** [reply]
Thanks
